src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoCommon.cs

```diff
@@ -304,41 +304,25 @@ public override IEnumerable CreateImmutableCollectionInstance(Type collectionTyp
         // Creates an IEnumerable<TRuntimePropertyType> and populates it with the items in the
         // sourceList argument then uses the delegateKey argument to identify the appropriate cached
         // CreateRange<TRuntimePropertyType> method to create and return the desired immutable collection type.
-        public override IDictionary CreateImmutableDictionaryInstance(Type collectionType, string delegateKey, IDictionary sourceDictionary, string jsonPath, JsonSerializerOptions options)
+        public override IDictionary CreateImmutableDictionaryInstance(ref ReadStack state, Type collectionType, string delegateKey, IDictionary sourceDictionary, JsonSerializerOptions options)
         {
             IDictionary collection = null;
 
             if (!options.TryGetCreateRangeDelegate(delegateKey, out ImmutableCollectionCreator creator) ||
                 !creator.CreateImmutableDictionary(sourceDictionary, out collection))
             {
-                ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(collectionType, jsonPath);
+                ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(collectionType, state.JsonPath);
             }
 
             return collection;
         }
 
-        private IEnumerable<TRuntimeProperty> CreateGenericTRuntimePropertyIEnumerable(IList sourceList)
-        {
-            foreach (object item in sourceList)
-            {
-                yield return (TRuntimeProperty)item;
-            }
-        }
-
         private IEnumerable<TDeclaredProperty> CreateGenericTDeclaredPropertyIEnumerable(IList sourceList)
         {
             foreach (object item in sourceList)
             {
                 yield return (TDeclaredProperty)item;
             }
         }
```

Comment:
Unrelated to your PR, but you might consider using Enumerable.Cast instead of this method:
https://github.com/dotnet/corefx/blob/bca549851e2b24eceb1968a4a8ebab5f2ba4f7f2/src/System.Linq/src/System/Linq/Cast.cs#L33-L55

Reply:
@layomia can you go through this code and update the casts per suggestion above?